index.bs

```diff
@@ -61,16 +61,31 @@ different storage mechanism with a different API for such files. The entry point
 
 A <dfn export id="entry">file system entry</dfn> is either a [=file entry=] or a [=directory entry=].
 
-Each [=/file system entry=] has an associated <dfn for="file system entry" id=entry-query-access>query access</dfn> algorithm, which takes "`read`"
-or "`readwrite`" <var ignore>mode</var> and returns a {{PermissionState}}. Unless specified
-otherwise it returns "{{PermissionState/denied}}". The algorithm is allowed to throw.
-
-Each [=/file system entry=] has an associated <dfn for="file system entry" id=entry-request-access>request access</dfn> algorithm, which takes
-"`read`" or "`readwrite`" <var ignore>mode</var> and returns a {{PermissionState}}. Unless specified
-otherwise it returns "{{PermissionState/denied}}". The algorithm is allowed to throw.
-
-Note: Implementations that only implement this specification and not dependent specifications do not
-need to bother implementing [=/file system entry=]'s [=file system entry/query access=] and [=file system entry/request access=].
+Each [=/file system entry=] has an associated
+<dfn for="file system entry" id=entry-query-access>query access</dfn>
+algorithm, which takes "`read`" or "`readwrite`" <var ignore>mode</var> and
+returns either a {{PermissionState}} or an [=exception/error name=] that must be
+listed in the [=error names table=].
+Unless specified otherwise it returns "{{PermissionState/denied}}".
+
+Each [=/file system entry=] has an associated
+<dfn for="file system entry" id=entry-request-access>request access</dfn>
+algorithm, which takes "`read`" or "`readwrite`" <var ignore>mode</var> and
+returns either a {{PermissionState}} or an [=exception/error name=] that must be
+listed in the [=error names table=].
+Unless specified otherwise it returns "{{PermissionState/denied}}".
+
+<p class=warning> Dependent specifications may consider this API a
+[=powerful feature=]. However, unlike other [=powerful features=] whose
+[=permission request algorithm=] may throw, [=/file system entry=]'s
+[=file system entry/query access=] and [=file system entry/request access=]
+algorithms must run [=in parallel=] on the [=file system queue=] and are
+therefore not allowed to throw. Instead, the caller is expected to [=/reject=]
+as appropriate should these algorithms return an [=exception/error name=].
```

Comment:
This really should go in WICG spec, but since we're doing something a bit funky it might be useful to have here, too?

Reply:
If they run in parallel the caller will also be in parallel. They wouldn't be able to immediately reject something.

But also I'm not a big fan of this design. Perhaps these should return a struct?

Reply:
> If they run in parallel the caller will also be in parallel. They wouldn't be able to immediately reject something.

Updated to say `the caller is expected to [=queue a storage task=] to [=/reject=], as appropriate`

... though while going through the whole spec I realized that there is one situation - the _asynchronous iterator initialization steps_ - where there is no promise to reject. See the other comment

> But also I'm not a big fan of this design. Perhaps these should return a struct?

Done. AFAICT a struct item cannot be "optional" (only algorithm parameters?) so "error name" is just empty most of the time